misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py

```diff
@@ -0,0 +1,419 @@
+import torch
+from torch import nn
+import torch.nn.functional as F
+import numpy as np
+
+
+class SpectralConv2d(nn.Module):
+    def __init__(self, ch, k, p, py, s, dil):
+        super().__init__()
+        self.ch = ch
+
+        self.ydim = nn.Conv2d(
+            ch, ch, (k, 1), padding=(
+                py, 0), stride=(
+                s, 1), dilation=(
+                dil, 1), bias=False)
+        self.xdim = nn.Conv2d(
+            ch, ch, (1, k), padding=(
+                0, p), stride=(
+                1, s), groups=ch, bias=False)
+        self.bias = nn.Conv2d(
+            ch, ch, (1, 1), padding=(
+                0, 0), stride=(
+                1, 1), groups=ch, bias=True)
+
+    def forward(self, x):
+        # do y x first follwed by the other spectal decompsed filters!
+
+        op = self.ydim(x)
+        #print(op.shape, x.shape)
+        op = self.xdim(op)
+        op = self.bias(op)
+
+        return op
+
+
+class SpectralConv2dInter(nn.Module):
+    def __init__(self, ch, k, pad, dil, wts_list=None, a=1):
+        super().__init__()
+        self.kernel = k + (k - 1) * (dil - 1)  # default dil = 1(no dilation)
+        self.pad = pad
+        index = torch.Tensor([i for i in range(0, self.kernel, dil)]).to(torch.int64)
+        z = torch.zeros(ch, ch, self.kernel, 1)
+        if wts_list is None:
+            print("Not implemented")
```

Comment:
If the code should not be executed further, then `NotImplementedError` should be raised. You later access a field of this object, so, I guess, `assert wts_list is not None, "some informative message if it is"` would be more applicable here.